tests/serialization/test_filter.py

```diff
@@ -66,3 +69,72 @@ def test_filter_array_like_categoricals(op, expected, cat_type):
     res = filter_array_like(ser, op, "B")
 
     assert all(res == expected)
+
+
+@pytest.mark.parametrize(
+    "value, filter_value",
+    [
+        (1, 1.0),
+        ("1", 1.0),
+        ("1", 1),
+        (1, "1"),
+        (datetime.date(2019, 1, 1), 1),
+        (datetime.datetime(2019, 1, 1), 1),
+        (datetime.datetime(2019, 1, 1), datetime.date(2019, 1, 1)),
+        (datetime.date(2019, 1, 1), datetime.datetime(2019, 1, 1)),
+    ],
+)
+@pytest.mark.parametrize("op", ["==", "!=", "<", "<=", ">", ">="])
+def test_raise_on_type(value, filter_value, op):
+    array_like = pd.Series([value])
+    with pytest.raises(TypeError, match="Unexpected type encountered."):
+        filter_array_like(array_like, op, filter_value, strict_date_types=True)
+
+
+def test_filter_df_from_predicates():
+    df = pd.DataFrame(
+        {
+            "A": range(10),
+            "B": ["A", "B"] * 5,
+            "C": pd.Series(["X", "Y"] * 5).astype("category"),
+            "D": [datetime.date(2019, 1, 1), datetime.date(2019, 1, 2)] * 5,
+            "E": [datetime.datetime(2019, 1, 1), datetime.datetime(2019, 1, 2)] * 5,
+        }
+    )
+
+    op = "=="
+    ix = 4
+    for col in df.columns:
+        value = df[col][ix]
+        predicates = [[(col, op, value)]]
+        actual = filter_df_from_predicates(df, predicates)
+        expected = df[df[col] == value]
+        pdt.assert_frame_equal(actual, expected)
+
+    op = "!="
+    for col in df.columns:
+        value = df[col][ix]
+        predicates = [[(col, op, value)]]
+        actual = filter_df_from_predicates(df, predicates)
+        expected = df[df[col] != value]
+        pdt.assert_frame_equal(actual, expected)
+
+    op = ">"
+    for col in df.columns:
+        value = df[col][ix]
+        predicates = [[(col, op, value)]]
+        actual = filter_df_from_predicates(df, predicates)
+        if pd.api.types.is_categorical(df[col]):
+            df[col] = df[col].astype(df[col].cat.as_ordered().dtype)
+        expected = df[df[col] > value]
+        pdt.assert_frame_equal(actual, expected, check_categorical=False)
+
+    op = "<"
+    for col in df.columns:
+        value = df[col][ix]
+        predicates = [[(col, op, value)]]
+        actual = filter_df_from_predicates(df, predicates)
+        if pd.api.types.is_categorical(df[col]):
+            df[col] = df[col].astype(df[col].cat.as_ordered().dtype)
+        expected = df[df[col] < value]
+        pdt.assert_frame_equal(actual, expected, check_categorical=False)
```

Comment:
Nitpick: can these for loops be compacted into a single nested loop? Don't know about the "correctness" of using `eval` but it could be suitable for this case ( as in `expected = eval(f"df[df[col] {op} value])"`)

Reply:
Also, just wondering, do `>=` and `<=` not apply here?

Reply:
I usually refrain from using eval/exec but in this case this is probably worth an exception.

> Also, just wondering, do >= and <= not apply here?

Yes but I didn't want to copy more code. The important aspect here was that ordered operators where tested. With the eval code I added them to them loop